applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp

```diff
@@ -0,0 +1,588 @@
+    }
+}
+
+/// <summary>
+/// Calculates the stiffness constants in all directions
+/// </summary>
+/// <param name="rVariables"></param>
+/// <param name="CurrentProcessInfo"></param>
+/// <param name="Geom"></param>
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::
+CalculateNodalStiffnessMatrix(NormalLysmerAbsorbingVariables& rVariables, const ProcessInfo& CurrentProcessInfo, Element::GeometryType& Geom)
+{
+    array_1d<double, 2> rStiffnessConstants;
+
+    // calculate rotation matrix
+    BoundedMatrix<double, TDim, TDim> rotationMatrix;
+    CalculateRotationMatrix(rotationMatrix, Geom);
+
+    const int local_perpendicular_direction = TDim - 1;
+
+
+    // calculate constant traction vector part
+    rStiffnessConstants[0] = rVariables.G / rVariables.virtual_thickness;
+    rStiffnessConstants[1] = rVariables.Ec / rVariables.virtual_thickness;
+
+
+    BoundedMatrix<double, TDim, TDim>             localKMatrix = ZeroMatrix(TDim, TDim);
+    BoundedMatrix<double, TDim, TDim>             auxLocalKMatrix = ZeroMatrix(TDim, TDim);
+
+    rVariables.KAbsMatrix = ZeroMatrix(TDim, TDim);
+
+    for (unsigned int idim = 0; idim < TDim; ++idim) {
+        localKMatrix(idim, idim) = rStiffnessConstants[0];
+    }
+    localKMatrix(local_perpendicular_direction, local_perpendicular_direction) = rStiffnessConstants[1];
+
+    auxLocalKMatrix = prod(localKMatrix, rotationMatrix);
+    rVariables.KAbsMatrix = prod(trans(rotationMatrix), auxLocalKMatrix);
+
+    for (unsigned int idim = 0; idim < TDim; ++idim) {
+        rVariables.KAbsMatrix(idim, idim) = abs(rVariables.KAbsMatrix(idim, idim));
+    }
+}
+
+/// <summary>
+/// This method gets the average of the variables of all the neighbour elements of the condition. 
+/// </summary>
+/// <param name="rVariables"></param>
+/// <param name="rCurrentProcessInfo"></param>
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::
+GetNeighbourElementVariables(
+    NormalLysmerAbsorbingVariables& rVariables, const ProcessInfo& rCurrentProcessInfo)
+{
+    
+    double rho_s = 0;
+    double rho_w = 0;
+    double E = 0;
+    double n = 0;
+    double nu = 0;
+    double rMeanDegreeOfSaturation = 0;
+
+    // get mean degree of saturation of all integration points in all neighbour elements
+    std::vector<double> SaturationVector;
+    
+    // get neighbour elements
+    auto neighbours = this->GetValue(NEIGHBOUR_ELEMENTS);
+    
+    int nElements = neighbours.size();
+    
+    // loop over all neighbour elements
+    int nValues = 0;
+    for (unsigned int i = 0; i < neighbours.size(); ++i) {
+        
+        auto const rPropNeighbour = neighbours[i].GetProperties();
+
+        rho_s = rho_s + rPropNeighbour[DENSITY_SOLID];
+        rho_w = rho_w + rPropNeighbour[DENSITY_WATER];
+        n = n + rPropNeighbour[POROSITY];
+        E = E + rPropNeighbour[YOUNG_MODULUS];
```

Comment:
Young modulus is not always available, for example for UDSMs/UMATs. Perhaps, you can get the pointer to the material model and get the stiffness matrix. Then you have access to G and Eoed.

Reply:
does this mean that the element in ConstitutiveMatrix(0,0) is always Eoed, and ConstitutiveMatrix(last_index,last_index) = G? or is it material or element dependent?

Reply:
At the moment, this is the case in all Geo-elements. You can find stress/strain indices of different elements in [https://github.com/KratosMultiphysics/Kratos/blob/master/applications/GeoMechanicsApplication/geo_mechanics_application_constants.h](url).

Reply:
Thanks, however, I see this as problematic, as to me it doesn't seem possible to recognize if a neighbour element is an interface element or not. Thus it would be impossible to either return an error, or get the parameters from the Interface element. Is it required to add an extra flag to the constitutive Law, e.g. ConstitutiveLaw::PLANE_STRAIN_INTERFACE_LAW  ?

Reply:
I see two options to solve this issue:

1. add two variables in GeoMech such as CONFINED_STIFFNESS (e_oed) and (SHEAR_STIFFNESS). Then we can calculate them in the element. In this way, the element will be responsible to give these variables.
2. Another possibility is to reorder stress/strain in the interface elements.

Reply:
Thanks for the suggestions, I think option 1 is most clear, I will implement that